src/main/java/org/elasticsearch/action/admin/indices/syncedflush/TransportSyncedFlushIndicesAction.java

```diff
@@ -0,0 +1,173 @@
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.action.admin.indices.syncedflush;
+
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.support.ActionFilters;
+import org.elasticsearch.action.support.HandledTransportAction;
+import org.elasticsearch.cluster.ClusterService;
+import org.elasticsearch.cluster.ClusterState;
+import org.elasticsearch.cluster.node.DiscoveryNode;
+import org.elasticsearch.cluster.routing.GroupShardsIterator;
+import org.elasticsearch.cluster.routing.ShardIterator;
+import org.elasticsearch.common.inject.Inject;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.util.concurrent.AbstractRunnable;
+import org.elasticsearch.common.util.concurrent.ConcurrentCollections;
+import org.elasticsearch.index.shard.ShardId;
+import org.elasticsearch.indices.SyncedFlushService;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.transport.*;
+
+import java.io.IOException;
+import java.util.Set;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Synced flush action
+ */
+public class TransportSyncedFlushIndicesAction extends HandledTransportAction<SyncedFlushIndicesRequest, SyncedFlushIndicesResponse> {
+
+
+    final private SyncedFlushService syncedFlushService;
+    final private ClusterService clusterService;
+    private final TransportService transportService;
+    public static final String INDICES_SYNCED_FLUSH_ACTION_NAME = "internal:indices/flush/synced/indices";
+
+    @Inject
+    public TransportSyncedFlushIndicesAction(Settings settings, ThreadPool threadPool, TransportService transportService, ActionFilters actionFilters, SyncedFlushService syncedFlushService, ClusterService clusterService) {
+        super(settings, SyncedFlushIndicesAction.NAME, threadPool, transportService, actionFilters, SyncedFlushIndicesRequest.class);
+        this.syncedFlushService = syncedFlushService;
+        this.clusterService = clusterService;
+        this.transportService = transportService;
+        transportService.registerRequestHandler(INDICES_SYNCED_FLUSH_ACTION_NAME, ShardSyncedFlushRequest.class, ThreadPool.Names.GENERIC, new ShardSyncedFlushHandler());
+    }
+
+
+    @Override
+    protected void doExecute(final SyncedFlushIndicesRequest request, final ActionListener<SyncedFlushIndicesResponse> listener) {
+        // we need to execute in a thread here because getSyncedFlushResults(request, state) blocks
+        threadPool.generic().execute(new AbstractRunnable() {
```

Comment:
I think this is not the way to go here - we should just make the service operate in an async fashion, I opened https://github.com/elastic/elasticsearch/pull/11111 for this